### PR TITLE
CLN: Simplify add_constant and add_trend

### PR DESCRIPTION
**Pull request by bashtage**
Simplified these two functions to handle NumPy arrays and recarrays as
well as pandas Series and DataFrames.  Removed bugs in add_trend.
Added complete test coverage for both functions.

### CODE REVIEW

**Comment by bashtage** [comment]
Found old issue related to these changes, so closed it too.

File: statsmodels/tools/tools.py [hunk 1]
```diff
@@ -233,107 +233,70 @@ def categorical(data, col=None, dictnames=False, drop=False, ):
...
+        data = np.column_stack((data, np.ones((data.shape[0], 1))))
         if prepend:
-            ones = np.ones((data.shape[0], 1), dtype=[('const', float)])
-            data = nprf.append_fields(ones, data.dtype.names,
-                                      [data[i] for i in data.dtype.names],
-                                      usemask=False, asrecarray=return_rec)
-        else:
-            data = nprf.append_fields(data, 'const', np.ones(data.shape[0]),
-                                      usemask=False, asrecarray=return_rec)
+            return np.roll(data, 1, 1)
```
Comment (josef-pkt) [inline statsmodels/tools/tools.py]: I think you should use the explicit `if prepend` paths.
`roll` will need to make a copy, or linalg will need to make a copy when it requires contiguous memory
also `column_stack` creates Fortran ordered arrays which should be better for the Fortran linalg calls, although we never verify this.

File: statsmodels/tsa/tsatools.py [hunk 2]
```diff
@@ -36,54 +37,62 @@ def add_trend(X, trend="c", prepend=False, has_constant='skip'):
...
-        else:
-            X = np.column_stack((trendarr, X))
+        trendarr = trendarr[:, 1]
+
+    if "c" in trend and np.any(np.ptp(np.asanyarray(X), axis=0) == 0):
+        if has_constant == 'raise':
+            raise ValueError("X already contains a constant")
+        elif has_constant == 'add':
+            pass
+        elif has_constant == 'skip' and trend == "ct":
```
Comment (josef-pkt) [inline statsmodels/tsa/tsatools.py]: this need to handle also "ctt", i.e. `trend in ["ct", "ctt"]` or  `trend_order > 0`

**Comment by josef-pkt** [comment]
I need to look at it again, but should be possible for 0.7

travis failure in `webuse` test, unrelated to this, and I need to fix that fast
`urllib.error.HTTPError: HTTP Error 503: Service Unavailable`

**Comment by bashtage** [comment]
`webuse` has issues.

**Comment by josef-pkt** [comment]
you mean the failing tests for `webuse` ?
They are disabled in master now. If you rebase, then they won't bother this PR anymore.

**Comment by josef-pkt** [comment]
Maybe this is simplifying too much.

AFAICS from one example that I tried, recarray -> DataFrame -> recarray doesn't roundtrip for string (but I didn't run the code of this PR)

```
>>> dt = np.dtype([('',int),('','<S4'), ('', np.float32), ('',np.float64)])
>>> xr = np.array([(1, 'aaa', 5.4, 1.2),(1, 'aaa', 5.4, 1.2)], dt).view(np.recarray)
>>> xr
rec.array([(1, 'aaa', 5.400000095367432, 1.2),
       (1, 'aaa', 5.400000095367432, 1.2)], 
      dtype=[('f0', '<i4'), ('f1', '|S4'), ('f2', '<f4'), ('f3', '<f8')])
>>> xrp = pd.DataFrame.from_records(xr)
>>> xrp.dtypes
f0      int32
f1     object
f2    float32
f3    float64
dtype: object
>>> xrp.to_records()
rec.array([(0L, 1, 'aaa', 5.400000095367432, 1.2),
       (1L, 1, 'aaa', 5.400000095367432, 1.2)], 
      dtype=[('index', '<i8'), ('f0', '<i4'), ('f1', '|O4'), ('f2', '<f4'), ('f3', '<f8')])
```

is there an option to roundtrip, i.e. not change the dtype?

Also I think the pandas detour will need to make some extra copies because the memory layout doesn't match up if there are different dtypes involved. However, given that these are mainly convenience function I don't think that's a blocker.

**If** we still have users that use recarrays, then changing the dtype will, most likely, break their code.

**Comment by bashtage** [comment]
This might not be popular, but I removed all of `add_constant` and made it a shallow wrapper for `add_trend`.  The code to handle mixed recarrays or DataFrames was somewhat complex, and having 2 copies did not seem like a good idea.  These functions were already extremely similar, and `add_trend` strictly nests `add_constant`.

**Comment by josef-pkt** [comment]
You can short-circuit our most common advertised use case: add_constant to ndarray
i.e. `if trend='c' and type(x) is np.ndarray: return column_stack(..)`

(I won't have time until later to look at it.)

**Comment by bashtage** [comment]
> You can short-circuit our most common advertised use case: add_constant to ndarray
> i.e. if trend='c' and type(x) is np.ndarray: return column_stack(..)

Unfortunately because of `has_constant` more checks would be needed. Almost all of the function is skipped when X is an `array` so I don't think this saves much in terms of time.

**Comment by josef-pkt** [comment]
I'd like to postpone this to 0.8. It has changes that I don't like to rush in just before a release, and I'm still not sure about some of the details, even though it looks good overall. For some changes, I don't know if they are innocent and acceptable or not.

**Comment by bashtage** [comment]
> is there an option to roundtrip, i.e. not change the dtype?

This works as it should in the PR.  `S4` stays `S4`.  Not that anyone should be ever using recarrays in 2016, except when essential to interface with C code (e.g. HDF5).

File: statsmodels/tsa/tsatools.py [hunk 3]
```diff
@@ -238,25 +273,20 @@ def detrend(x, order=1, axis=0):
...
-    x = np.asarray(x)
+    """
+    if x.ndim == 2 and int(axis) == 1:
+        x = x.T
+    elif x.ndim > 2:
+        raise NotImplementedError('x.ndim > 2 is not implemented until it is needed')
+    # could use a polynomial, but this should work also with 2d x, but maybe not yet
     nobs = x.shape[0]
-    if order == 0:
-        return x - np.expand_dims(x.mean(axis), axis)
```
Comment (josef-pkt) [inline statsmodels/tsa/tsatools.py]: I would like to keep this. There is no reason to go through linalg to demean.

File: statsmodels/tools/tools.py [hunk 4]
```diff
@@ -233,108 +232,37 @@ def categorical(data, col=None, dictnames=False, drop=False, ):
...
+    -----
+    When the input is recarray or a pandas Series or DataFrame, the added
+    column's name is 'const'.
     """
-    if _is_using_pandas(data, None):
-        # work on a copy
-        return _pandas_add_constant(data.copy(), prepend, has_constant)
-    else:
-        data = np.asarray(data)
-    if not data.dtype.names:
```
Comment (josef-pkt) [inline statsmodels/tools/tools.py]: I think we should also handle the plain numpy array case here. 
It's the most common case, and it's a long way to track the code to where it is done in the current changes, when all we need is essentially a np.column_stack, even with a bit of code duplication.

I would also check for constants only if has_constant != 'add'

**Comment by bashtage** [comment]
I added the special case, even though it takes quite a bit of code (to handle exception cases).

File: statsmodels/tools/tools.py [hunk 5]
```diff
@@ -233,108 +232,57 @@ def categorical(data, col=None, dictnames=False, drop=False, ):
...
+        return add_trend(data, trend='c', prepend=prepend, has_constant=has_constant)
+
+    # Special case for NumPy
+    x = np.asanyarray(data)
+    if x.ndim == 1:
+        x = x[:,None]
+    elif x.ndim > 2:
+        raise ValueError('Only implementd 2-dimensional arrays')
+
+    is_nonzero_const = np.ptp(x, axis=0) == 0
```
Comment (josef-pkt) [inline statsmodels/tools/tools.py]: this part could be skipped, using `if has_constant != 'add'`

File: statsmodels/tsa/tsatools.py [hunk 6]
```diff
@@ -238,25 +273,25 @@ def detrend(x, order=1, axis=0):
...
+        raise NotImplementedError('x.ndim > 2 is not implemented until it is needed')
 
-
-    '''
-    x = np.asarray(x)
     nobs = x.shape[0]
     if order == 0:
-        return x - np.expand_dims(x.mean(axis), axis)
+        # Special case demean
+        resid = x - x.mean(axis=0)
```
Comment (josef-pkt) [inline statsmodels/tsa/tsatools.py]: This ignores axis argument.

Comment (josef-pkt) [inline statsmodels/tsa/tsatools.py]: No, it uses axis argument because internally x is transposed if needed

**Comment by josef-pkt** [comment]
also has now merge conflict.

one small speedup, otherwise looks good to me and ready to merge

File: statsmodels/tsa/tsatools.py (hunk 6 above)
Comment (bashtage) [inline statsmodels/tsa/tsatools.py]: Yes

**Comment by bashtage** [comment]
Rebased and squashed a few pep8 style warnings

**Comment by josef-pkt** [comment]
this failed on all travis runs

`TypeError: add_trend() got an unexpected keyword argument 'X'`

**Comment by bashtage** [comment]
Yes - semi-zealous pep8 cleaning missed some keyword useage in testing for
a required argument.

**Comment by josef-pkt** [comment]
this should also close #2656 #2659   supersedes/closes #2665

**Comment by bashtage** [comment]
Green.

**Comment by josef-pkt** [comment]
merged

@bashtage  Thank you

File: statsmodels/tools/tools.py [hunk 7]
```diff
@@ -233,108 +232,57 @@ def categorical(data, col=None, dictnames=False, drop=False, ):
...
+    is_nonzero_const &= np.all(x != 0.0, axis=0)
```
Comment (josef-pkt) [inline statsmodels/tools/tools.py]: another small speedup: if ptp == 0, then all values are the same and we only need to check 1 element of x: x[0] != 0

**Pull request merged**